qa/L0_model_update/instance_update_test.py

```diff
@@ -0,0 +1,310 @@
+        # Load model
+        update_instance_group("{\ncount: 4\nkind: KIND_CPU\n}")
+        self.__triton.load_model("instance_init_del")
+        self.assertEqual(get_initialize_count(), 4)
+        self.assertEqual(get_finalize_count(), 0)
+        self.__triton.infer("instance_init_del", self.__get_inputs((2,)))
+        # Update batching and instance
+        enable_batching()
+        update_instance_group("{\ncount: 2\nkind: KIND_CPU\n}")
+        self.__triton.load_model("instance_init_del")
+        time.sleep(10)  # wait for unload to complete
+        self.assertEqual(get_initialize_count(), 6)
+        self.assertEqual(get_finalize_count(), 4)
+        self.__triton.infer("instance_init_del", self.__get_inputs((1, 1)))
+        # Unload model
+        self.__triton.unload_model("instance_init_del")
+        time.sleep(10)  # wait for unload to complete
+        self.assertEqual(get_initialize_count(), 6)
+        self.assertEqual(get_finalize_count(), 6)
+        with self.assertRaises(InferenceServerException):
+            self.__triton.infer("instance_init_del", self.__get_inputs((1, 4)))
+
+    def test_update_while_inferencing(self):
+        # Load model
+        update_instance_group("{\ncount: 1\nkind: KIND_CPU\n}")
+        self.__triton.load_model("instance_init_del")
+        self.assertEqual(get_initialize_count(), 1)
+        self.assertEqual(get_finalize_count(), 0)
+        self.__triton.infer("instance_init_del", self.__get_inputs((2,)))
+        # Add 1 instance while inferencing
+        set_infer_delay(10)
+        update_instance_group("{\ncount: 2\nkind: KIND_CPU\n}")
+        with concurrent.futures.ThreadPoolExecutor() as pool:
+            infer_start_time = time.time()
+            infer_thread = pool.submit(self.__triton.infer, "instance_init_del",
+                                       self.__get_inputs((1,)))
+            time.sleep(2)  # make sure inference has started
+            update_start_time = time.time()
+            update_thread = pool.submit(self.__triton.load_model,
+                                        "instance_init_del")
+            update_thread.result()
+            update_end_time = time.time()
+            infer_thread.result()
+            infer_end_time = time.time()
+        infer_time = infer_end_time - infer_start_time
+        update_time = update_end_time - update_start_time
+        # Adding a new instance does not depend on existing instances, so the
+        # ongoing inference should not block the update.
+        self.assertGreaterEqual(infer_time, 10.0, "Invalid infer time")
+        self.assertLess(update_time, 5.0, "Update blocked by infer")
+        self.assertEqual(get_initialize_count(), 2)
+        self.assertEqual(get_finalize_count(), 0)
+        self.__triton.infer("instance_init_del", self.__get_inputs((8,)))
+        # Unload model
+        self.__triton.unload_model("instance_init_del")
+        time.sleep(10)  # wait for unload to complete
```

Comment:
Instead of sleeping for 10 sec, can we poll the finalize_count every second for an entire period of 10s?
This will allow us to return earlier than 10sec.

Reply:
Updated: [Refactor instance update test](https://github.com/triton-inference-server/server/pull/5635/commits/f012b0474ac4b1465e432e72f3c4f24cb5cb6b3c)